localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp

```diff
@@ -0,0 +1,233 @@
+      input.yaw = base.yaw + dist_normal_(engine_) * coeff * yaw_stddev_;
+
+      // fixed angle
+      input.roll = fix_angle(input.roll);
+      input.pitch = fix_angle(input.pitch);
+      input.yaw = fix_angle(input.yaw);
+    }
+    const double score = acquisition_function(input);
+    if (score > best_score) {
+      best_score = score;
+      best_input = input;
+    }
+  }
+  return best_input;
+}
+
+double TreeStructuredParzenEstimator::acquisition_function(const Input & input)
+{
+  const int64_t n = trials_.size();
+
+  // The upper KDE and the lower KDE are calculated respectively, and the ratio is the score of the
+  // acquisition function.
+  std::vector<double> upper_logs;
+  std::vector<double> lower_logs;
+
+  // Scott's rule
+  const double coeff_upper = BASE_STDDEV_COEFF * std::pow(good_num_, -1.0 / 10);
+  const double coeff_lower = BASE_STDDEV_COEFF * std::pow(n - good_num_, -1.0 / 10);
+  const Input sigma_upper = base_stddev_ * coeff_upper;
+  const Input sigma_lower = base_stddev_ * coeff_lower;
+
+  std::vector<double> weights;
+  constexpr double PRIOR_WEIGHT = 0.5;
+  double upper_sum = PRIOR_WEIGHT;
+  double lower_sum = 0.0;
+  for (int64_t i = 0; i < n; i++) {
+    const double v = trials_[i].score - MIN_GOOD_SCORE;
+    const double w = std::min(std::abs(v), 1.0);
+    if (i < good_num_) {
+      weights.push_back(w);
+      upper_sum += w;
+    } else {
+      weights.push_back(w);
+      lower_sum += w;
+    }
+  }
+
+  for (int64_t i = 0; i < n; i++) {
+    if (i < good_num_) {
+      const double log_p = log_gaussian_pdf(input, trials_[i].input, sigma_upper);
+      const double w = weights[i] / upper_sum;
+      const double log_w = std::log(w);
+      upper_logs.push_back(log_p + log_w);
+    } else {
+      const double log_p = log_gaussian_pdf(input, trials_[i].input, sigma_lower);
+      const double w = weights[i] / lower_sum;
+      const double log_w = std::log(w);
+      lower_logs.push_back(log_p + log_w);
+    }
+  }
+
+  // prior
+  // Only yaw has a large variance in order to have a pseudo-uniform distribution.
+  const Input zeros{};
+  const Input sigma{x_stddev_, y_stddev_, z_stddev_, roll_stddev_, pitch_stddev_, M_PI * 2};
+  const double log_p = log_gaussian_pdf(input, zeros, sigma);
+  const double log_w = std::log(PRIOR_WEIGHT / upper_sum);
+  upper_logs.push_back(log_p + log_w);
+
+  auto log_sum_exp = [](const std::vector<double> & log_vec) {
+    const double max = *std::max_element(log_vec.begin(), log_vec.end());
+    double sum = 0.0;
+    for (const double log_v : log_vec) {
+      sum += std::exp(log_v - max);
+    }
+    return max + std::log(sum);
+  };
+
+  const double upper = log_sum_exp(upper_logs);
+  const double lower = log_sum_exp(lower_logs);
+  const double r = upper - lower;
+  return r;
+}
+
+double TreeStructuredParzenEstimator::log_gaussian_pdf(
+  const Input & input, const Input & mu, const Input & sigma)
```

Comment:
Please add `const` or make this function independent from TreeStructuredParzenEstimator (fix_angle function as well)
```suggestion
double TreeStructuredParzenEstimator::log_gaussian_pdf(
  const Input & input, const Input & mu, const Input & sigma) const
```

Reply:
Fixed when changing to generic TPE.